Feature: suwayomi Service Widget

Robonau opened the pull request:
## Proposed change
add a [Suwayomi](https://github.com/Suwayomi/Suwayomi-Server) Service widget
Closes #2972

![image](https://github.com/user-attachments/assets/f5782051-89bc-4d17-a2fa-f7076122d577)

I had to use a custom proxy because i needed GraphQL variables for variable category selection.
in the case of `category: all` the response from Suwayomi would look like.
```json
{
  "data": {
    "downloadCount": {
      "totalCount": 46428
    },
    "nonDownloadCount": {
      "totalCount": 46428
    },
    "readCount": {
      "totalCount": 31345
    },
    "UnreadCount": {
      "totalCount": 33770
    }
  }
}
```
~200B

in the case of `category: {number}`
```json
{
  "data": {
    "category": {
      "mangas": {
        "nodes": [
          {
            "chapters": {
              "nodes": [
                {
                  "isRead": false,
                  "isDownloaded": true
                },
                {
                  "isRead": true,
                  "isDownloaded": true
                },
                ...
              ]
            }
          },
          ...
        ]
      }
    }
  }
}
```
~6Kib (for a massive library)
you can't filter chapters by what category they are in or filter chapters from the mangas or category queries
so this was the best solution.

The response sent to the homepage client would be
```json
[
  {
    "count": 46428,
    "label": "suwayomi.download"
  },
  {
    "count": 46428,
    "label": "suwayomi.nondownload"
  },
  {
    "count": 31345,
    "label": "suwayomi.read"
  },
  {
    "count": 33770,
    "label": "suwayomi.unread"
  }
]
```
~400B
regardless of category value

<!--
Please include a summary of the change. Screenshots and/or videos can also be helpful if appropriate.

*** Please see the development guidelines for new widgets: https://gethomepage.dev/more/development/#service-widget-guidelines
*** If you do not follow these guidelines your PR will likely be closed without review.


New service widgets should include example(s) of relevant API output as well as updates to the docs for the new widget.
-->

## Type of change

<!--
What type of change does your PR introduce to Homepage?
-->

- [x] New service widget
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation only
- [ ] Other (please explain)

## Checklist:

- [x] If applicable, I have added corresponding documentation changes.
- [x] If applicable, I have reviewed the [feature](https://gethomepage.dev/more/development/#new-feature-guidelines) and / or [service widget guidelines](https://gethomepage.dev/more/development/#service-widget-guidelines).
- [x] I have checked that all code style checks pass using [pre-commit hooks](https://gethomepage.dev/more/development/#code-formatting-with-pre-commit-hooks) and [linting checks](https://gethomepage.dev/more/development/#code-linting).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

shamoon reviewed (changes requested):
I’m going to mark this as draft as it does not actually seem ready to merge. There are tons of comments and overall the code seems a bit messy. I’m happy to provide more specifics if you need but at the moment I’d like to wait on reviewing this until it’s a bit more polished

Robonau commented:
I've cleaned up the non jsdoc comments.
fixed a couple issues i found.
moved a bunch of code in to functions outside of the default function in the proxy.
refactored the makeBody and extractCounts functions.

I think this is as clean as its gonna get without any feedback.

shamoon commented:
Thanks ill take a look when i have a second

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,269 @@
+  nondownload: {
+    condition: (c) => !c.isDownloaded,
+    gqlCondition: "isDownloaded: false",
+  },
+  read: { condition: (c) => c.isRead, gqlCondition: "isRead: true" },
+  unread: { condition: (c) => !c.isRead, gqlCondition: "isRead: false" },
+  downloadedread: { condition: (c) => c.isDownloaded && c.isRead, gqlCondition: "isDownloaded: true, isRead: true" },
```
```suggestion
  },
  read: {
    condition: (c) => c.isRead,
    gqlCondition: "isRead: true"
  },
  unread: {
    condition: (c) => !c.isRead,
    gqlCondition: "isRead: false"
  },
  downloadedread: {
    condition: (c) => c.isDownloaded && c.isRead,
    gqlCondition: "isDownloaded: true, isRead: true"
  },
```

Just more legible

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,269 @@
+function makeFields(Fields = []) {
+  let fields = Fields ?? [];
+  if (fields.length === 0) {
+    fields = ["download", "nonDownload", "read", "unRead"];
```
There is a lot of inconsistency of the camelCase vs all lowercase for fields. Please double-check

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,269 @@
+ * @param {string[]|null} Fields
+ * @returns {string[]}
+ */
+function makeFields(Fields = []) {
```
Please take the updated code from compontent.jsx and again, dont bother making a separate method. Its obvious what it does, its pretty simple and is only used once.

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,269 @@
+ * @param {widget} widget
+ * @returns {{ "Content-Type": string, Authorization?: string }}
+ */
+function makeHeaders(widget) {
```
Again, no reason for a separate method here.

Robonau commented on src/widgets/suwayomi/component.jsx:
```diff
@@ -0,0 +1,68 @@
+  if (suwayomiError) {
+    return <Container service={service} error={suwayomiError} />;
+  }
+
```
fair i was just copied it from the proxy since it needed to do the same thing.

It did handle it with the nullish coalescing to [].

shamoon commented on src/widgets/suwayomi/component.jsx:
```diff
@@ -0,0 +1,41 @@
+    if (!widget.fields || widget.fields.length === 0) {
+      widget.fields = ["download", "nondownload", "read", "unread"];
+    } else if (widget.fields.length > 4) {
+      widget.fields = widget.fields.slice(0, 4);
+      widget.fields = widget.fields.map((field) => field.toLowerCase());
```
```suggestion
      widget.fields = widget.fields.slice(0, 4);
```

Dont do the lowercase, it just allows user errors to go un-noticed.

shamoon commented on docs/widgets/services/suwayomi.md:
```diff
@@ -0,0 +1,20 @@
+
+Learn more about [Suwayomi](https://github.com/Suwayomi/Suwayomi-Server).
+
+Allowed fields:["download", "nondownload", "read", "unread", "downloadedread", "downloadedunread", "nondownloadedread", "nondownloadedunread"]
```
```suggestion
Allowed fields: ["download", "nondownload", "read", "unread", "downloadedread", "downloadedunread", "nondownloadedread", "nondownloadedunread"]
```

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,178 @@
+    widget.fields = ["download", "nondownload", "read", "unread"];
+  } else if (widget.fields.length > 4) {
+    widget.fields = widget.fields.slice(0, 4);
+    widget.fields = widget.fields.map((field) => field.toLowerCase());
```
```suggestion
```

shamoon commented on src/widgets/suwayomi/proxy.js:
```diff
@@ -0,0 +1,178 @@
+    return res.status(status).send({ error: { message: "Error getting data. body: %s, data: %s", body, data } });
+  }
+
+  const responseJSON = JSON.parse(data);
+
+  const returnData = extractCounts(responseJSON, widget.fields);
```
```suggestion
  const returnData = extractCounts(JSON.parse(data), widget.fields);
```

DRY

shamoon reviewed (approved):
<img width="750" alt="Screenshot 2024-11-11 at 5 52 40 PM" src="https://github.com/user-attachments/assets/23ff07f6-bf28-4d9c-9524-696ff47c232c">

Great thanks.
